numba/np/arrayobj.py

```diff
@@ -1655,7 +1655,8 @@ def fancy_setslice(context, builder, sig, args, index_types, indices):
                                       builder.icmp_signed('!=', u, v))
 
         with builder.if_then(shape_error, likely=False):
-            msg = "cannot assign slice from input of different size"
+            msg = (f"cannot assign slice of shape {src_shapes} from input of"
+                   f" shape {index_shape}")
```

Comment:
I don't think this will print `src_shapes` or `index_shape` in the exception message. `context.call_conv.return_user_exc` will add a constant string (see below) to the LLVM module at compile-time, but both variables are only known at runtime. It will actually print the following:

```python
from numba import njit
import numpy as np

@njit
def bar(a):
    a[:] = np.arange(3)
    return a

a = np.arange(5)
print(bar(a))

ValueError: cannot assign slice of shape [<ir.ExtractValue '.374' of type 'i64', opname 'extractvalue', operands [<ir.InsertValue '.373' of type '[1 x i64]', opname 'insertvalue', operands [<ir.Constant type='[1 x i64]' value=<llvmlite.ir.values._Undefined object at 0x12a41b460>>, <ir.LoadInstr '.372' of type 'i64', opname 'load', operands [<ir.GEPInstr '.371' of type 'i64*', opname 'getelementptr', operands [<ir.GEPInstr '.370' of type '[1 x i64]*', opname 'getelementptr', operands [<ir.AllocaInstr '.341' of type '{i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64]}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=5>]>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>]>]>] from input of shape (<ir.SelectInstr '.323' of type 'i64', opname 'select', operands [<ir.SelectInstr '.322' of type 'i1', opname 'select', operands [<ir.ICMPInstr '.313' of type 'i1', opname 'icmp', operands [<ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>, <ir.Constant type='i64' value=0>]>, <ir.ICMPInstr '.320' of type 'i1', opname 'icmp', operands [<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=0>]>, <ir.ICMPInstr '.321' of type 'i1', opname 'icmp', operands [<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=0>]>]>, <ir.Constant type='i64' value=0>, <ir.Instruction '.319' of type 'i64', opname 'add', operands (<ir.Constant type='i64' value=1>, <ir.Instruction '.318' of type 'i64', opname 'sdiv', operands (<ir.SelectInstr '.317' of type 'i64', opname 'select', operands [<ir.ICMPInstr '.313' of type 'i1', opname 'icmp', operands [<ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>, <ir.Constant type='i64' value=0>]>, <ir.Instruction '.316' of type 'i64', opname 'add', operands (<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=1>)>, <ir.Instruction '.315' of type 'i64', opname 'sub', operands (<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=1>)>]>, <ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>)>)>]>,)
```

To fix that you can use `context.call_conv.return_dynamic_user_exc`:
```python
with builder.if_then(shape_error, likely=False):
    # **Wrongly** assuming that len(src_shapes) == len(index_shape) == 1
    msg = ("cannot assign shape of shape ", src_shapes[0],
            "from input of shape ", index_shape[0])
    nb_types = (None, types.int64, None, types.int64)
    context.call_conv.return_dynamic_user_exc(builder, ValueError,
                                                msg, nb_types)
```

Which will give you the following error message:

```python
$ python repro.py
Traceback (most recent call last):
  File "/Users/guilhermeleobas/git/numba/repro.py", line 117, in <module>
    print(bar(a))
ValueError: ('cannot assign shape of shape ', 3, 'from input of shape ', 5)
```

This is not exactly what you would expect (no string interpolation) but circumvent the issue of raising dynamic exceptions. I'll create an issue for tracking string interpolation inside an exception.

Reply:
Ahh, thanks for clarifying :) Ill see if I can get to it soon enough

Reply:
Let me know if you need any help. I think the interface for `call_conv.return_dynamic_user_exc` could be improved.

Reply:
@guilhermeleobas For the case above here, we don't know that it is the first index that causes the trouble. Is there a type for tubles / lists in the lowering I can use here?

I tried `Tuple`, `List`, and also flattening `*[types.int32 for _ in src_shapes]`
but it does not seem to work
```python
 with builder.if_then(shape_error, likely=False):
            msg = ("cannot assign slice of shape ", src_shapes,
                   "from input of shape ", index_shape)

            nb_types = (None, types.Tuple, None, types.Tuple)
            context.call_conv.return_dynamic_user_exc(builder, ValueError, msg,
                                                       nb_types)
```

Reply:
I think I found an ok compromise:

```python
# repro.py
from numba import njit
import numpy as np

@njit
def bar(a):
    a[:] = np.arange(3)
    return a
a = np.arange(12).reshape((3, 2, 2))

print(bar(a))
# ValueError: ('cannot assign slice of shape', 3, 2, 3, 'from input of shape', 3, 2, 2)
```

Reply:
Nice, thank you! 
Implemented -- although I have not been able to test the Sequence - part of this flow, i don't know how to reproduce this error (or even how to construct a numba Sequence ....)